src/server/frontend_wayland/layer_shell_v1.cpp

```diff
@@ -159,7 +160,16 @@ void mf::LayerSurfaceV1::set_keyboard_interactivity(uint32_t keyboard_interactiv
 
 void mf::LayerSurfaceV1::get_popup(struct wl_resource* popup)
 {
-    (void)popup;
+    auto popup_window_role =
+        static_cast<WindowWlSurfaceRole*>(
+            static_cast<XdgPopupStable*>(
+                static_cast<wayland::XdgPopup*>(
+                    wl_resource_get_user_data(popup))));
```

Comment:
That looks horrid. Would this work?
```C++
auto const popup_window_role = dynamic_cast<WindowWlSurfaceRole*>(wayland::XdgPopup::from(popup);
```

Reply:
Because `XdgPopupStable` inherits from both `wayland::XdgPopup` and `WindowWlSurfaceRole`, so you can't cast directly between the two unrelated base classes. Stand by for a `XdgPopupStable::from` implementation to clear this up.

Reply:
But casting directly between two unrelated base classes is exactly what `dynamic_cast` is good for!?

Reply:
Oh, cool. I assumed `dynamic_cast` could only handle down casts, but now I see it somehow has enough runtime info to pull that off. Anyway, the new system with `XdgPopupStable::from()` is still better IMO.